db/clientlist_virtual_reader.cc

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright (C) 2019 ScyllaDB
+ */
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "db/clientlist_virtual_reader.hh"
+
+#include "api/api.hh"
+#include "db/system_keyspace.hh"
+#include "db/timeout_clock.hh"
+#include "dht/i_partitioner.hh"
+#include "mutation_fragment.hh"
+#include "range.hh"
+#include "service/storage_service.hh"
+#include "transport/server.hh"
+
+#include <algorithm>
+#include <unordered_map>
+#include <vector>
+
+namespace db::clientlist {
+
+namespace {
+class clients_reducer {
+    std::vector<system_keyspace::client_data> _val;
+public:
+    future<> operator()(std::vector<system_keyspace::client_data>&& rhs) {
+        // Move the smaller vector into the bigger one
+        if (_val.size() < rhs.size()) {
+            _val.swap(rhs);
+        }
+        _val.reserve(_val.size() + rhs.size());
+        _val.insert(_val.end(), std::make_move_iterator(rhs.begin()), std::make_move_iterator(rhs.end()));
+        return make_ready_future<>();
+    }
+
+    std::vector<system_keyspace::client_data> get() && {
+        return std::move(_val);
+    }
+};
+
+class ck_comparator {
```

Comment:
Suggestion: define this at the point of usage, i.e. inside the filter_and_sort_by_port function. It will be easier to understand why this thing's needed without jumping around the code as much.